### PR TITLE
Adds metric for tracking merges initiated from the branch list

### LINKED ISSUE #6414: [Metrics] Merges from branch dropdown
**What question are we trying to answer?**

We'd like to understand if we made merging more discoverable by including a way to merge in the branch dropdown?

**What metrics do we need to help answer that?**

- [ ] Absolute number of merges initiated via the new button in the branch dropdown.

screen shot 2018-12-13 at 1 46 10 pm

cc: @iAmWillShepherd @ampinsk @telliott27

### PR DESCRIPTION
**Pull request by iAmWillShepherd**
Fixes #6414

⚠️ ⚠️ ⚠️ Whoever reviews this should also review the related PRs ⚠️ ⚠️ ⚠️ 

This PR adds the `mergesIniatedFromBranchList ` metric. This PR also introduces a refactor. The previous way we recorded merges required us to keep track of state across multiple places. I've chosen to move all merge metric incrementation to a single function. I can now do this because I introduced some extra state, `initiatedBy`, to the `Popup` type with `type` equal to ` PopupType.MergeBranch`

`initiatedBy` is a property that keeps track of the place where the merge event took place. For instance, if I click _Merge_ dialog is opened because I clicked _Merge Into Current Branch..._, `initiatedBy` will be set to `MergeSource.MergeIntoCurrentBranchMenuItem`. Once execution reaches the ` _mergeBranch`, I can inspect that property and increment the right thing.

I introduced this change because the alternative was to keep some form of state in `AppStore` because some of the ways to get to the _Merge_ dialog require multiple clicks and keeping short-term state like that can be tedious. 

FWIW, I'm open to naming suggestions for the `MergeSource` type and the `initiatedBy` field on the Popup and in function parameter lists.

**Related tasks**
- [x] Open issue in github/analytics - https://github.com/github/analytics/issues/808
- [x] Update central - https://github.com/github/central/pull/423
- [x] Update marketing site - https://github.com/github/desktop.github.com/pull/122/commits/8c4eaa320895f7a381a153383164237457ea816e (updated an open PR instead of creating new one)
- [x] Update metrics doc - https://github.com/desktop/desktop/pull/6424/commits/616f4044b99298d59e37ca7988c1fe6fffabcd92

### CODE REVIEW

**Comment by iAmWillShepherd** [comment]
# Items
Pax G-13
Select White Fire

**Merge from compare tab**
```ts
/** The number of times a merge is initiated in the `compare` sidebar */
mergesInitiatedFromComparison
```

**Merge master from top-level menu**
```ts
/** The number of times the `Branch -> Update From Default Branch` menu item is used */
  readonly updateFromDefaultBranchMenuCount: number
```

**Merge some branch into current branch from top-level menu**
```ts
  /** The number of times the `Branch -> Merge Into Current Branch` menu item is used */
  readonly mergeIntoCurrentBranchMenuCount: number
```

**Merge from NDDB notification**
```ts
  /** The number of times the user merges from the diverged branch notification merge CTA button */
  readonly divergingBranchBannerInitatedMerge: number
```

```ts
  /**
   * The number of times the user merges from the compare view after getting to that state
   * from the diverged branch notification compare CTA button
   */
  readonly divergingBranchBannerInfluencedMerge: number
```

**Merge from branch list**
```ts
  /** The number of times the user iniated a merge from within the branch dropdown */
  readonly mergesIniatedFromBranchList: number
```

 :point_up:  was removed  since we to only want to increment out counter if they did the merge, not clicked the button.  The `divergingBranchBannerInitatedMerge` metric keeps track if they did the merge.

**Comment by billygriffin** [comment]
> ☝️ was removed since we to only want to increment out counter if they did the merge, not clicked the button. The divergingBranchBannerInitatedMerge metric keeps track if they did the merge.

@iAmWillShepherd Which thing are you saying was removed here?

Separately, I'm a little concerned that `divergingBranchBannerInfluencedMerge` and `mergesInitiatedFromComparison` might be double-counting toward the total. We want the percentages for each of these ways of initiating a merge, but we also want a reliable total merges initiated for @telliott27 and @scottypate to be able to analyze. That total number is what we use to evaluate, for example, what percentage of merges resulted in merge conflicts.

You may be accounting for all that here, just wanted to make sure. 😄

**Comment by iAmWillShepherd** [comment]
@billygriffin I was intending to remove the metric, `divergingBranchBannerInfluencedMerge`. I think this should be fine because we don't use that metric in any calculations that I know of, but I'll have to confirm with @telliott27. I will not be merging this today, so I'll think on this over the weekend.

**Comment by tierninho** [comment]
Testing is done ✅for these:

- [x] `mergesInitiatedFromComparison`
- [x] `updateFromDefaultBranchMenuCount`
- [x] `mergeIntoCurrentBranchMenuCount` 
- [x] `divergingBranchBannerInitatedMerge`
- [x] `divergingBranchBannerInfluencedMerge`
- [x] `mergesIniatedFromBranchList`

mac, 1.5.1-beta4, 2e78eeee98c450a20078ed18a1de828b1535cad4

File: docs/process/metrics.md [hunk 1]
```diff
@@ -59,9 +64,5 @@ These are general metrics about feature usage and specific feature behaviors. Th
 | `repoWithIndicatorClicked` | The numbers of times a repo with indicators is clicked on repo list view. | To understand usage patterns around the repository indicators feature. |
 | `repoWithoutIndicatorClicked` | The numbers of times a repo without indicators is clicked on repo list view.  | To understand usage patterns around the repository indicators feature. |
 | `unattributedCommits` | The number of commits that will go unattributed to GitHub users. | To understand how frequently commits in GitHub Desktop are unattributed and how highly we should prioritize design for those instances |
-| `updateFromDefaultBranchMenuCount` | The number of times the `Branch -> Update From Default Branch` menu item is used. | To understand usage patterns around the compare branches feature. |
```
Comment (iAmWillShepherd) [inline docs/process/metrics.md]: These were not actually removed. This list became unsorted, so I resorted it.

File: app/src/models/merge.ts [hunk 2]
```diff
@@ -36,3 +36,12 @@ export interface IMergeUnsupported {
 }
 
 export type MergeResult = IMergeSuccess | IMergeError | IMergeUnsupported
+
+export enum MergeSource {
+  BranchDropDown,
+  MergeIntoCurrentBranchMenuItem,
+  UpdateBranchMenuItem,
+  Compare,
```
Comment (shiftkey) [inline app/src/models/merge.ts]: :art: could this be renamed to `CompareTab` - it feels less contextual at first glance compared to the others

File: app/src/lib/stores/app-store.ts [hunk 3]
```diff
@@ -3083,6 +3086,30 @@ export class AppStore extends TypedBaseStore<IAppState> {
     return this._refreshRepository(repository)
   }
 
+  private incrementMergeMetric(source: MergeSource) {
+    switch (source) {
+      // merge from within compare tab
```
Comment (shiftkey) [inline app/src/lib/stores/app-store.ts]: :art: these comments could live on the enum itself as human-friendly jsdoc comments

File: app/src/models/merge.ts [hunk 4]
```diff
@@ -36,3 +36,12 @@ export interface IMergeUnsupported {
 }
 
 export type MergeResult = IMergeSuccess | IMergeError | IMergeUnsupported
+
+export enum MergeSource {
+  BranchDropDown,
```
Comment (shiftkey) [inline app/src/models/merge.ts]: :art: convert this to a string enum to simplify the generated JS [[link](https://www.typescriptlang.org/play/#src=export%20enum%20MergeSourceNumber%20%7B%0D%0A%20%20BranchDropDown%2C%0D%0A%20%20MergeIntoCurrentBranchMenuItem%2C%0D%0A%20%20UpdateBranchMenuItem%2C%0D%0A%20%20Compare%2C%0D%0A%20%20DivergingBranchNotification%2C%0D%0A%20%20NewCommitsBanner%2C%0D%0A%7D%0D%0A%0D%0Aexport%20enum%20MergeSourceString%20%7B%0D%0A%20%20BranchDropDown%20%3D%20'BranchDropDown'%2C%0D%0A%20%20MergeIntoCurrentBranchMenuItem%20%3D%20'UpdateBranchMenuItem'%2C%0D%0A%20%20UpdateBranchMenuItem%20%3D%20'UpdateBranchMenuItem'%2C%0D%0A%20%20Compare%20%3D%20'Compare'%2C%0D%0A%20%20DivergingBranchNotification%20%3D%20'DivergingBranchNotification'%2C%0D%0A%20%20NewCommitsBanner%20%3D%20'NewCommitsBanner'%2C%0D%0A%7D%0D%0A)]

**Comment by billygriffin** [comment]
@shiftkey @iAmWillShepherd I removed the milestone here. @scottypate and Will and I are discussing metrics more broadly tomorrow and I think this is probably just worth touching out before it gets in. Thanks!

**Comment by billygriffin** [comment]
@iAmWillShepherd What do you think about this now that we're looking at our metrics quality more generally?

**Comment by shiftkey** [comment]
@iAmWillShepherd is this still in scope for `1.6.3`? It doesn't seemed to have moved much since the last review, and there are some conflicts that need to be resolved.

**Comment by shiftkey** [comment]
Dropping this from the milestone because the linked issue #6414 does not have a milestone attached, and it's not clear to me if this is close to another round of review.

**Comment by iAmWillShepherd** [comment]
Closing this PR until we have bandwidth to resume out event metrics overhaul.

**Pull request closed without merging**